### DISCUSSION TITLE
Native TypeScript support

### DISCUSSION THREAD

**Discussion by sidharthachatterjee** [gatsbyjs/gatsby#18983] [Ideas / Feature Requests] [answered]
TypeScript is fantastic and has been gaining a lot of popularity. 

One _oft_ requested feature in Gatsby has been native TypeScript support<sup>1</sup>. We see no reason to _not_ support it out of the box so let's discuss how to make this happen. 

### Note
1. This will not change current DX for non-TypeScript users. 
1. This will result in deprecating [gatsby-plugin-typescript](https://www.gatsbyjs.org/packages/gatsby-plugin-typescript/) and fix some of the caveats it has (documented in its README)

### To do 
- [x] Support TypeScript in GraphQL query results? 
- [x] Support ts and tsx files out of the box
- [ ] Linting support in CLI (if using TypeScript)
- [x] Potentially support ts for `gatsby-*` files 

### Foot notes
1. https://twitter.com/chatsidhartha/status/1155820333095505920

### Related issues
https://github.com/gatsbyjs/gatsby/issues/16310,https://github.com/gatsbyjs/gatsby/issues/17934, https://github.com/gatsbyjs/gatsby/issues/15277, https://github.com/gatsbyjs/gatsby/issues/14185, https://github.com/gatsbyjs/gatsby/issues/13436, https://github.com/gatsbyjs/gatsby/issues/1457, https://github.com/gatsbyjs/gatsby/issues/10452

### Related PRs

https://github.com/gatsbyjs/gatsby/pull/16952
https://github.com/gatsbyjs/gatsby/pull/18721

### Related plugins
[gatsby-plugin-typescript](https://www.gatsbyjs.org/packages/gatsby-plugin-typescript/)
[gatsby-plugin-typescript-checker](https://www.gatsbyjs.org/packages/gatsby-plugin-typescript-checker/)
[gatsby-plugin-generate-typings](https://www.gatsbyjs.org/packages/gatsby-plugin-generate-typings/)
[gatsby-plugin-typegen](https://www.gatsbyjs.org/packages/gatsby-plugin-typegen/)
[gatsby-plugin-graphql-codegen](https://www.gatsbyjs.org/packages/gatsby-plugin-graphql-codegen/)

**Comment by anikethsaha** [comment]
Is it supporting through re writing the whole in `.ts` ? 
If so, I guess it will be a huge amount of work with a package of this much big codebase.

**Comment by sidharthachatterjee** [comment]
@anikethsaha This wouldn't involve rewriting the gatsby code base. We're discussing the possibility of adding native TypeScript support for userland code.

**Comment by maecapozzi** [comment]
I think this would be really helpful for those of us who write TypeScript and ❤️Gatsby. I'm particularly excited about these two potential changes: 

## Support `ts` and `tsx` files out of the box
I like that this won't be a breaking change, and `js` and `jsx` files will still be supported, so nothing is taken away from JS users. 

I'm curious about how you're considering supporting this? Would it be as simple as converting the Gatsby codebase to TypeScript with an `allowJs` set to true in the `tsconfig.json`?  

## Potentially support `ts` for `gatsby-*` files
At Harry's, we write some complicated logic in our `gatsby-*` files, and often talk about how we wish we could write them in TypeScript. 


## Other Questions
* What version of TS are you hoping to support? 
* Are there potential challenges with keeping Gatsby's GraphQL type system in sync with TypeScript types? I'm thinking particularly about `gatsby-*` files.

**Comment by anikethsaha** [comment]
> We're discussing the possibility of adding native TypeScript support for userland code.

Support using `.d.ts` files ? like using type defs ?

**Comment by ghost** [comment]
maby we can make a `gatsby-ts.js` file with the TypeScript config in it to the base starter

**Comment by NickyMeuleman** [comment]
> Potentially support ts for gatsby-* files

Yessssss!

Been trying out https://github.com/gatsbyjs/gatsby/issues/1457#issuecomment-381405638 and https://gist.github.com/clarkdave/53cc050fa58d9a70418f8a76982dd6c8 as a "good enough" solution for now.

Official support would be absolutely wonderful.
The big pull of TS for me is the autocomplete and in-editor documentation. That would definitely make logic in `gatsby-*` files (like schema customization) much easier.

**Comment by orta** [comment]
The DTS support for Gatsby's API is pretty solid already also, though I think it might be worth thinking about how to document it for folks too, [for example](https://github.com/and-hyde/example/blob/master/example/src/lib/createPages.ts#L1-L27) I got really far using indexed types like this:

```ts
import { GatsbyNode } from "gatsby"
import * as path from "path"

export const createPages: GatsbyNode["createPages"] 
  = async ({ graphql, actions }) => {

  const { createPage } = actions

  const blogPost = path.resolve(`./src/templates/blog-post.tsx`)
  const result = await graphql(`
    {
      allMarkdownRemark(sort: { fields: [frontmatter___date], order: DESC }, limit: 1000) {
        edges {
          node {
            fields {
              slug
            }
...
```

**Comment by maecapozzi** [comment]
I'd love to understand more of the details of how the Core Team is considering implementing this! Seems like there are three potential approaches: 

1. TypeScript through the TypeScript compiler
2. TypeScript through `babel-typescript`
3. TS with JSDoc

About a year ago, I researched this a lot to set up a lerna-managed monorepo using TypeScript. We landed on using `@babel/preset-typescript`, although we do have a type-checking script that uses the typescript compiler (`tsc --noEmit`). 

The only place we've had trouble with `@babel/preset-typescript` is when we wanted to use a package with decorators.

**Comment by anikethsaha** [comment]
> maby we can make a `gatsby-ts.js` file with the TypeScript config in it to the base starter

This sounds interesting. 

I dont think typescript compiler will fir here. 
Rather it seems the babel typescript thing can work here. Anything which can be opt-in and out easily can be considered

**Comment by ghost** [comment]
> > maby we can make a `gatsby-ts.js` file with the TypeScript config in it to the base starter
> 
> This sounds interesting.
> 
> I dont think typescript compiler will fir here.
> Rather it seems the babel typescript thing can work here. Anything which can be opt-in and out easily can be considered

Thats what im saying seeing as we already have `gatsby-browser.js` / `gatsby-ssr.js` ext.. and can choose to opt out of using those via just deleting the file i dont see why we cant do something like `gatsby-ts.js` or similar

**Comment by ghost** [comment]
For added example there can be a `gatsby-ts.js` file with the config in it and if user deletes the file in the build stage or development stage whichever in the CMD there can be a message something like `We see you have chosen to opt-out of using Typescript for this project` and then go along with the process of building the web app ext.. and there can be something base lvl detecting if the user deleted the file or not example

```javascript
if(!gatsby-ts.js){
  // force stop of running anything typescript related
  console.log('We see you have chosen to opt-out of using Typescript for this project');
}else {
  // run typescript file here
}
```

just a thought and an example is all this ovi is a WIP but yea

**Comment by maecapozzi** [comment]
> For added example there can be a `gatsby-ts.js` file with the config in it and if user deletes the file in the build stage or development stage whichever in the CMD there can be a message something like `We see you have chosen to opt-out of using Typescript for this project` and then go along with the process of building the web app ext.. and there can be something base lvl detecting if the user deleted the file or not example
> 
> ```js
> if(!gatsby-ts.js){
>   // force stop of running anything typescript related
>   console.log('We see you have chosen to opt-out of using Typescript for this project');
> }else {
>   // run typescript file here
> }
> ```
> 
> just a thought and an example is all this ovi is a WIP but yea

With this example, is it possible to go back to opting into TypeScript, or is this like "ejecting?"

**Comment by ghost** [comment]
@maecapozzi what do you mean im not understanding sorry

**Comment by msallent** [comment]
YES! Please.

I love Gatsby and I love TypeScript. I even created a starter that used `gatsby-plugin-typescript` but I decided to go back to JS as I didn’t like it at all.

Native support would be awesome, specially having code not compile if there are typing errors instead of having to run tsc manually every time.

**Comment by MartinRosenberg** [comment]
@msallent I recommend trying [`gatsby-plugin-ts`](https://www.gatsbyjs.org/packages/gatsby-plugin-ts/) in the meantime, which actually integrates TypeScript into the flow, and even generates GraphQL types for you. I've been building [a starter](https://github.com/Collabsalot/gatsby-starter) with it, and it's been a pretty good experience so far.

**Comment by hasparus** [comment]
quick question: **Is TypeScript type generation from GraphQL a part of this issue?**

Configuring codegen to get access to types generated by Gatsby is a little bit to hard at this moment IMHO.

I had to set up both `gatsby-plugin-codegen` and `graphql-codegen` task in the package.json 😅
I didn't try gatsby-plugin-ts mentioned by @MartinRosenberg though.
At a first look, I think it doesn't generate types for nodes for usage in `gatsby-node.js` though.

```js
exports.onCreateNode = ({ node, getNode, actions: { createNodeField } }) => {
  if (node.internal.type === "Mdx") {
    /** @type {import("./__generated__/global").Mdx} */
    const mdxNode = node;
```

---

> maby we can make a gatsby-ts.js file with the TypeScript config in it to the base starter

Wouldn't `tsconfig.json` be enough? Next.js detects TSConfig and fills it with default values (but not in an obtrusive way like CRA).

**Comment by mdluo** [comment]
Original reply: https://github.com/gatsbyjs/gatsby/issues/1457#issuecomment-554641107

There was already some solutions about writing `gatsby-node` in TypeScript in that issue, my reply is more about dealing with GraphQL.

Use [gatsby-plugin-typegen](https://github.com/cometkim/gatsby-plugin-typegen) to generate types from GraphQL schema so that there will be:

- [x] GraphQL syntax highlight, with [VSCode GraphQL extension](https://marketplace.visualstudio.com/items?itemName=Prisma.vscode-graphql)
- [x] GraphQL eslint checking with [eslint-plugin-graphql](https://github.com/apollographql/eslint-plugin-graphql)
- [x] Type definitions of the query result

In both gatsby-node configs and tsx components.

Example: https://github.com/mdluo/blog-gatsby/commit/68eb2693a67fbafc869fd55272d04bc3e96f4021

GraphQL syntax highlight and eslint:
![image](https://user-images.githubusercontent.com/4597409/68994360-2ae95f80-08bd-11ea-93fb-f58e758488d8.png)

Types of query result:
![image](https://user-images.githubusercontent.com/4597409/68994380-52402c80-08bd-11ea-97e8-ffb26a0b9312.png)

**Comment by xndyz** [comment]
FYI all the codgen plugins for generating typings for Gatsby graphql queries are broken. Native support to get TS working with Gatbsy static queries and files would be amazing ❤️

**Comment by LekoArts** [comment]
Would appreciate if you could define „broken“ 😊 Thanks!

**Comment by rubas** [comment]
> FYI all the codgen plugins for generating typings for Gatsby graphql queries are broken. Native support to get TS working with Gatbsy static queries and files would be amazing ❤️

It's not like you would need a magic gatsby-plugin for this. 

Just use `@graphql-codegen`and run a watcher. 
You could use `gatsby-plugin-extract-schema` to have extract the current schema during the build process to a file - or just point codegen to `http://localhost:8000/___graphql`.

**The problem** is that `graphql(..)` command in createPages/Nodes is not picked up by codegen. We solve this with a "shadow file", that has a copy of this queries but call them via `graphql from gatsby`. It's annoying to maintain this shadow file but it works nicely to generate all types.

**Comment by me4502** [comment]
Just a reason for adding TypeScript support to gatsby-config etc, would be codesharing. Eg, a fair few plugins have options that require some more complex code, and this is often then duplicated across TypeScript and JavaScript files. It being in TypeScript means you can include it from the rest of the codebase, without having to maintain a second JS version (or lose type safety all together)

**Comment by orta** [comment]
Just to link everything, I wrote on the Gatsby blog a little bit about TS + Gatbsy: https://www.gatsbyjs.org/blog/2020-01-23-why-typescript-chose-gatsby/

**Comment by cometkim** [comment]
Announcing the gatsby-plugin-typegen v1 is finally released!!

https://github.com/cometkim/gatsby-plugin-typegen

- Re-implemented based on gatsby internal. This plugin is now using gatsby actions directly instead of using 3rd party tools like chokidar and graphql-toolkit.
- BOOST performance!!
- Switch to yarn 2 (berry). It is much faster and helps this plugin resolve all problems about dependency resolution.
- Support Flow as an option.
- Support emitting multiple schema files.
- Support emitting plugin documents.
- Quiet log until you set --verbose
- Add more integration use cases into README

This plugin helps using static/page queries with type-safety out-of-box.

Hope you enjoy :)

**Comment by Js-Brecht** [comment]
I'm really looking forward to having OOTB Typescript support with Gatsby.

Until then:

The beta version of `gatsby-plugin-ts-config` has just been released.  It will add `ts-node` as an interpreter during build, for translating your configuration files.

https://github.com/Js-Brecht/gatsby-plugin-ts-config

Now you can write _**all**_ of your config files in Typescript!

It definitely needs a lot of testing out in the real world, so if you feel like giving it a whirl, and you run into problems, please report them!

https://github.com/Js-Brecht/gatsby-plugin-ts-config/issues

**Comment by ascorbic** [comment]
@cometkim @Js-Brecht Awesome work. I wonder how much of your work would be worth bringing into core, and how much should stay a plugin.

**Comment by cometkim** [comment]
@ascorbic Typegen plugin's emitting and code-fixing features can be unnecessary overhead for users who don't want to.

And this is already being handled almost using Gatsby API, there's no big problem yet to use it from the outside. (I think it's rather decoupled well) But moving the package into the Gatsby monorepo would be ok.

The next big milestone I think is enhanced code-fixing and schema customization. I wrote every type definitions for Gatsby API every time I use them such as sore actions and resolver context, especially typing the resolver context would be pretty big. It's desirable that these type definitions are eventually merged into the Gatsby core.

**Comment by Js-Brecht** [comment]
@ascorbic I'm all for bringing something similar to `gatsby-plugin-ts-config` into the core, but that's based off of my personal opinion.  I feel that the best utility to use for transpiling Typescript is Typescript itself; and that's what TS-Node does.  Babel is great as a post-ts processing step... but for the TS itself 🤷‍♂.  I've always run into too much of a headache when using Babel for transpiling Typescript.

However, this would definitely need to be something a user can opt in to, so there would need to be a pre-build hook, or some other method, for configuring Typescript support aside from, and before, the usual `gatsby-*` processing.  `ts-node` is great, and it makes it really easy to use Typescript's language service as an interpreter, but it does add overhead that could potentially have a negative impact on performance.  However, I'm sure that if this were integrated into the core, a few things could be done to tune the performance, and the team at `ts-node` is actively pursuing performance enhancement (https://github.com/TypeStrong/ts-node/issues/754).<sup>1</sup>

Obviously, a great deal of work would have to be done to integrate `ts-node`, or something similar, into the build chain.  Maybe some inspiration could be taken from `gatsby-plugin-ts-config`, but since it's basically just a workaround to make the plugin system work for this task, it would need to be something rather different to fit effectively in the core.

---

<sup>1</sup> _That particular issue doesn't have a lot of bearing on Gatsby, because Gatsby configuration files wouldn't have nearly the surface area as a set of tests, and the execution method is different; but I think performance as a whole will probably benefit from their work there.  In the core, `ts-node` can also be tuned to only process very specific files (more specifically, their dependencies), which should, theoretically, increase performance too._

**Comment by orta** [comment]
FWIW: I think it's fine for the built-in support for TypeScript to come through the babel plugin, it's how it's done in react/react-native/next/jest etc.

Doing this means only one or two extra (small) deps for Gatsby itself vs the 8mb monolith of adding TypeScript and then additional tooling like ts-node. Given that Gatsby already has the babel pipeline setup for JSX, it's not a big distance then between JS and TS users.

**Comment by Js-Brecht** [comment]
And it will soon have babel support by default.  Just need to finish the docs.

**Comment by liam-brooks** [comment]
> And it will soon have babel support by default. Just need to finish the docs.

It would be great!

**Comment by cometkim** [comment]
Hi folks, I'm thinking of the next of my plugin [gatsby-plugin-typegen](https://github.com/cometkim/gatsby-plugin-typegen).

The plugin, unlike other plugins, doesn't run the codegen watcher, it integrates deeper into the Gatsby lifecycle and runs codegen directly, so [it's much faster and smoother and it doesn't hurt development experience such as hot reloading](https://github.com/nibtime/gatsby-firebase-pwa-template/issues/1).

I am aware of some bugs in it but it still works.

The downside of this approach is that it is sensitive to changes in the core behavior.

I love Gatsby's content sourcing and enjoy the power of it and fragment colocation. The last piece is definitely a strong typing support, so I hope we can support this.

I'm also thinking about the typings generated by the experimental File system API. The easiest way to support all the cases in the user node is to make it from the core, or Make it from third-party requires very uncertain reverse engineering and code analysis.

I'd like to implement the next version in a reliable way as possible. I'm wondering if there is a possibility to port my plugin to the Gatsby core.

**Comment by LekoArts** [comment]
@cometkim Have you talked to @acao ? They are working on a VSCode GraphQL extension and a gatsby plugin that would generate typings if I remember correctly.

**Comment by LekoArts** [comment]
Hi!

We're in the first phase of "TypeScript All The Things" for Gatsby. You'll be able to write all files with TypeScript. Please go to https://github.com/gatsbyjs/gatsby/discussions/34613 and give it a try. In the next phase we'll look at having better support for GraphQL codegen. You can watch the RFC discussion category for this or our changelog + Twitter to see when this lands.

**Comment by LekoArts** [comment, marked as answer]
Hey, we're now in the second phase of "TypeScript All The Things"!

Since Gatsby v4.9 you're able to use TS in `gatsby-*` files 🎉  https://www.gatsbyjs.com/docs/reference/release-notes/v4.9/

Soon you'll be able to use built-in typegen in Gatsby! Give this a try: [RFC: GraphQL TypeScript Generation](https://github.com/gatsbyjs/gatsby/discussions/35420)
